---
title: "FinalProject"
author: "LNass"
date: "December 3, 2017"
output: html_document
---

```{r echo=FALSE, message=FALSE}
library(tidyverse)
library(gridExtra)
library(nlme)
library(car)
library(multcomp)
library(vegan)
library(gplots)
library(ape)
library(devtools)
#install_github("ggbiplot", "vqv")
library(ggbiplot)
library(dendextend)
library(pander)
library(data.table)
```


# Exploratory Data Analysis

There are 4 data files and 3 supplemental research papers

```{r}
lipids = read.table("Gacu_lipids_data.tsv", header = T, row.names = 1)
rnaseq = read.table("Gacu_rnaseq_data.tsv", header = T, row.names = 1)
sex = read.table("Gacu_sex_data.tsv", header = T, row.names = 1)
survival = read.table("Gacu_survival_data.tsv", header = T, row.names = 1)
```

First let's look at each data set individually

## Lipids

This file contains HPLC measurements of triacylglyceride (TAG) concentration in ug/mg for whole bodies
of 14 dpf stickleback larvae. Again, effects of population (Boot Lake or Rabbit Slough) and microbiota
treatment (GF or CV) are of interest. 

Here we have two explanatory variables, Population, Sex, and Microbiota, and one response variable, TAG concentration. In this dataset we have a total of 80 observations.

```{r}
#First get an idea of lipid distribution
ggplot(data	=	lipids)	+	
  geom_histogram(mapping	= aes(x	=	TAGs),	binwidth	= .0005) + ggtitle("Distribution of Overall Lipids")

#Shapiro test for normality
shapiro.test(lipids$TAGs)
#It seems here that the general TAGs data is far from a normal distribution, doesn't obviously look like any other distribution either. So instead we'll look at the different explanatory variables and see if those provide a normal distribution.

CV = subset(lipids, Microbiota == 'CV')
GF = subset(lipids, Microbiota == 'GF')

shapiro.test(CV$TAGs)
shapiro.test(GF$TAGs)

p1 = ggplot(data	=	CV)	+	
  geom_histogram(mapping	= aes(x	=	TAGs)) + ggtitle("CV Microbiota Lipid Distribution")

p2 = ggplot(data	=	GF)	+	
  geom_histogram(mapping	= aes(x	=	TAGs)) + ggtitle("GF Microbiota Lipid Distribution")

RS = subset(lipids, Population == 'RS')
Bt = subset(lipids, Population == 'Bt')

shapiro.test(RS$TAGs)
shapiro.test(Bt$TAGs)


p3 = ggplot(data	=	RS)	+	
  geom_histogram(mapping	= aes(x	=	TAGs)) + ggtitle("RS Population Lipid Distribution")

p4 = ggplot(data	=	Bt)	+	
  geom_histogram(mapping	= aes(x	=	TAGs)) + ggtitle("Bt Population Lipid Distribution")

grid.arrange(p1,p2,p3,p4, ncol = 2)
```
The GF Microbiota treatment and the Bt Population both show normal distributions. However the other two treatments do not. While the individual explanatory variables show overall better fits than all lipids, our analysis needs to account the lack of unanimous normal distributions.

## RNA-seq

This file contains transcript abundance measures from RNA-seq data for individual 14 dpf fish intestines.
Data for 300 genes expressed in the gut are included here. The units of the expression values
are gene-specific read counts per 1 million total reads aligned. These can be understood as normalized
expression values. Population, Microbiota, and Sex are of interest here. 

In this data set we have 3 explanatory variables: Population, Microbiota and Sex, and 300 response variables which are all the expressed gene transcript levels. We have 48 observations in this data set.

```{r}
means = lapply(rnaseq, mean)
means = as.data.frame(means)
means = t(means)

#Here we can see our lowest expressed gene is ENSGACG00000016101 and our highest is ENSGACG00000003794

p1 = ggplot(data	=	rnaseq)	+	
  geom_histogram(mapping	= aes(x	=	ENSGACG00000016101)) + ggtitle("Lowest Overall Mean Expression Level")

p2 = ggplot(data	=	rnaseq)	+	
  geom_histogram(mapping	= aes(x	=	ENSGACG00000003794)) + ggtitle("Highest Overall Mean Expression Level")

grid.arrange(p1,p2, ncol = 2)
#Note that values are normalized
```

## Sex

This file contains information about whether each experimental fish was female or male. Sex determination
in stickleback is genetic (XY), and these fish were sexed by PCR amplification of a sex-specific
marker. Also included is the population (Boot Lake or Rabbit Slough) from which each fish was derived,
and the microbiota treatment (GF or CV) each fish received. 

This file serves as an archive for the genders of the fish. All the variables are explanatory and the observations all match the lipid dataset, so the analysis can be performed in tandem with lipids.

## Survival

This file contains information about whether each experimental fish survived to 14 days post fertilization,
or died. Also included is the population (Boot Lake or Rabbit Slough) from which each fish was derived,
and the microbiota treatment (GF or CV) each fish received.

We have two explanatory variables, Population and Microbiota, and one categorical response variable: survival. For this dataset we have 160 observations.

```{r}
#Here we'll visualize our survival data
#ggplot(data	=	survival)	+	
#  geom_bar(mapping	= aes(x	=	Survived)) + ggtitle("Overall Survival")

RS = subset(survival, Population == "RS")
p1 = ggplot(data	=	RS)	+	
  geom_bar(mapping	= aes(x	=	Survived, fill = Survived)) + ggtitle("RS Population Survival")

Bt = subset(survival, Population == "Bt")
p2 = ggplot(data	=	Bt)	+	
  geom_bar(mapping	= aes(x	=	Survived, fill = Survived)) + ggtitle("Bt Population Survival")

GF = subset(survival, Microbiota == "GF")
p3 = ggplot(data	=	GF)	+	
  geom_bar(mapping	= aes(x	=	Survived, fill = Survived)) + ggtitle("GF Microbiota Survival")

CV = subset(survival, Microbiota == "CV")
p4 = ggplot(data	=	CV)	+	
  geom_bar(mapping	= aes(x	=	Survived, fill = Survived)) + ggtitle("CV Microbiota Survival")

grid.arrange(p1,p2,p3,p4, ncol = 2)
```

# Data Analysis

## Lipids

With regards to the lipids data, we are interested in the effect the different explanatory variables have on lipid concentration. The best analysis to conduct for this would be a multifactorial ANOVA. In this case we have 3 factors, all fixed. However One of our factors (Sex) is not balanced (45 vs. 35 samples), so we must proceed as a Model I Unbalanced Multifactor ANOVA.

First we have to see if we meet assumptions.
```{r}
par(mfrow = c(2, 2))
boxplot(lipids$TAGs ~ lipids$Population)
boxplot(lipids$TAGs ~ lipids$Microbiota)
boxplot(lipids$TAGs ~ lipids$Sex)
boxplot(lipids$TAGs ~ lipids$Population * lipids$Microbiota * lipids$Sex)

```
Our first assumption is that the samples are normally distributed, while there is some discrepancy here most evident in the final interaction and to a lesser degree in the population and microbiota factors, it should be within tolerable ANOVA paramters and to be kept in mind for further hypothesis tests. Our second assumption is no relationship between mean and variance and it seems to be met with confidence in all plots except our final interaction. Lastly we have independence which was adressed in the experimental design. A transformation was tested however seemed unnecessary for this data.

We know our design is unbalanced, so a type III sums of squared will be used.

```{r}
#Set our contrasts
contrasts(lipids$Population) = contr.helmert
contrasts(lipids$Microbiota) = contr.helmert
contrasts(lipids$Sex) = contr.helmert

#lipids_AOV = aov(TAGs ~ Population * Microbiota * Sex, data = lipids)
#lipids_AOV

#plot(lipids_AOV)

#Check our replications to confirm unbalanced design
replications(TAGs ~ Population * Microbiota * Sex, data = lipids)

#Perform type III multifactorial anova
lipid_aov = Anova(lm(TAGs ~ Population * Microbiota * Sex, data = lipids),type = "3")
lipid_aov
lipid_aov = aov(TAGs ~ Population * Microbiota, data = lipids)

#Perform our post-hoc test
TukeyHSD(lipid_aov)

# Summarise Interaction with a bar plot
lipid_int1 = aggregate(lipids$TAGs,
                  by = list(pop = lipids$Population, micro = lipids$Microbiota),
                  FUN = function(x) c(mean = mean(x), sd = sd(x),
                                      n = length(x)))
lipid_int2 <- do.call(data.frame, lipid_int1)

lipid_int2$se <- lipid_int2$x.sd / sqrt(lipid_int2$x.n)
colnames(lipid_int2) <- c("Population", "Microbiota", "mean", "sd", "n", "se")
lipid_int2$names <- c(paste(lipid_int2$Population, "Population /",
                        lipid_int2$Microbiota, " Microbiota"))

dodge <- position_dodge(width = 0.9)
limits <- aes(ymax = lipid_int2$mean + lipid_int2$se,
              ymin = lipid_int2$mean - lipid_int2$se)

ggplot(data = lipid_int2, aes(x = names, y = mean, fill = names)) +
  geom_bar(stat = "identity", position = dodge) +
  geom_errorbar(limits, position = dodge, width = 0.25) +
  theme(axis.text.x=element_blank(), axis.ticks.x=element_blank(),
        axis.title.x=element_blank())+
  guides(fill=guide_legend(title="Factor Combinations")) + ggtitle("Lipid Concentration") + ylab("Mean TAG Conc (ug/mg)")
```
We see a significant effect with regards to the population the fish originated from (F-value and p-value: 1.1296e+05 < 2.2e-16), and the microbiota they were subjected to (F-value and p-value: 5.2452e+01 3.941e-10). Additionally we see a sinfigicance int he interaction effect between the two variables, 2.1263e+02 < 2.2e-16.

## RNA-Seq

Here we begin to analyize the RNA-seq data. This will be done with a combination of clustering and principle component analysis.
```{r}

#clus.dis1 <- vegdist(rnaseq)
test3 = rnaseq
test3$Population = NULL
test3$Microbiota = NULL
test3$Sex = NULL

bypop = test3

#Look at sample names as populations to evaluate patterns in tree

bypop <- setattr(bypop, "row.names", c("Bt1", "Bt2", "Bt3", "Bt4", "Bt5", "Bt6", "Bt7", "Bt8", "Bt9", "Bt10", "Bt11", "Bt12", "Bt13", "Bt14", "Bt15", "Bt16", "Bt17", "Bt18", "Bt19", "Bt20", "Bt21", "Bt22", "Bt23", "Bt24", "RS1", "RS2", "RS3", "RS4", "RS5", "RS6", "RS7", "RS8", "RS9", "RS10", "RS11", "RS12", "RS13", "RS14", "RS15", "RS16", "RS17", "RS18", "RS19", "RS20", "RS21", "RS22", "RS23", "RS24"))

clus.dis1 <- vegdist(bypop)
cluster_complete <- hclust(clus.dis1, "complete")
plot(cluster_complete, main = "RNA-Seq samples by Microbiota")

##############

#Change the sample names in order to evaluate any patterns in our tree
bymicro = arrange(rnaseq, Microbiota)
bymicro <- setattr(bymicro, "row.names", c("GF1", "GF2", "GF3", "GF4", "GF5", "GF6", "GF7", "GF8", "GF9", "GF10", "GF11", "GF12", "GF13", "GF14", "GF15", "GF16", "GF17", "GF18", "GF19", "GF20", "GF21", "GF22", "GF23", "GF24", "CV1", "CV2", "CV3", "CV4", "CV5", "CV6", "CV7", "CV8", "CV9", "CV10", "CV11", "CV12", "CV13", "CV14", "CV15", "CV16", "CV17", "CV18", "CV19", "CV20", "CV21", "CV22", "CV23", "CV24"))
bymicro$Population = NULL
bymicro$Microbiota = NULL
bymicro$Sex = NULL


clus.dis1 <- vegdist(bymicro)
cluster_complete <- hclust(clus.dis1, "complete")
plot(cluster_complete, main = "RNA-Seq samples by Population", xlab = "")

#We tried clustering by both population and microbiota however saw no distinct patterns.
```

```{r}

#Here we create a heatmap of our RNA data to look at the different levels of gene expression
#test3 = t(test3)

#Perform additional exploratory analysis on the data, trying different methods

#clus.dis1 <- vegdist(test3)
#clus.dis2 <- vegdist(decostand(test3, "hell"), "euclidean")
#cluster <- hclust(clus.dis1, "single")
#plot(cluster)

#cluster_complete <- hclust(clus.dis1, "complete")
#plot(cluster_complete)

#Create a heatmap to see our transcript expressions
heatmap.2(as.matrix(log10(test3+1)),
 scale="row", key=T,
 density.info="none", trace="none",cexCol=0.9, labRow=NA, dendrogram = "both", labCol = T, main = "Transcript Expression Levels")

```
Our heatmap shows the samples on the y axes and our genes on the x axis.


Now we'll also investigate our sample with PCA

```{r}
########################################### PCA Analysis
# log transform 
log.ir <- log(test3+1)
ir.pop <- rnaseq$Population
ir.mic <- rnaseq$Microbiota
ir.sex <- rnaseq$Sex

# apply PCA - scale. = TRUE is highly 
# advisable, but default is FALSE. 
ir.pca <- prcomp(log.ir,
                 center = TRUE,
                 scale. = TRUE) 

plot(ir.pca, type = "l")
summary(ir.pca)
#Our plot here shows us how well the components explain the variance, we will focus on the first 2 though the 3rd will also be examined.

# summary method
#summary(ir.pca)
p1 = NULL
p1 <- ggbiplot(ir.pca, obs.scale = 1, var.scale = 1, 
              groups = ir.pop, ellipse = TRUE, 
              circle = TRUE, varname.size = 0, var.axes = FALSE)
p1 <- p1 + scale_color_discrete(name = '')
p1 <- p1 + theme(legend.direction = 'horizontal', 
               legend.position = 'top')


p2 = NULL
p2 <- ggbiplot(ir.pca, obs.scale = 1, var.scale = 1, 
              groups = ir.mic, ellipse = TRUE, 
              circle = TRUE, varname.size = 0, var.axes = FALSE)
p2 <- p2 + scale_color_discrete(name = '')
p2 <- p2 + theme(legend.direction = 'horizontal', 
               legend.position = 'top')

p3 = NULL
p3 <- ggbiplot(ir.pca, obs.scale = 1, var.scale = 1, 
              groups = ir.sex, ellipse = TRUE, 
              circle = TRUE, varname.size = 0, var.axes = FALSE)
p3 <- p3 + scale_color_discrete(name = '')
p3 <- p3 + theme(legend.direction = 'horizontal', 
               legend.position = 'top')

grid.arrange(p1,p2,p3, ncol = 3)
```

We performed Principle Component Analysis on our RNA data set then attempted to group by population, microbiota, and sex. We see the majority of the variance can be explained by population, as seen in the leftmost figure. There is a clear population distinction. Some clustering by treatment is also observed though not as distinctive as population. Sex showed no clustering at all. Groups were also examined using PC3 and PC4 (8% nd 7% respectively), however no clear clustering was observed.

## Sex

As all our sex data is categorical, what we examine here are the number of samples belonging to each factor.
```{r}
nrow(subset(subset(subset(sex, Sex == "M"), Population == "Bt"),Microbiota = "CV"))
nrow(subset(subset(subset(sex, Sex == "M"), Population == "Bt"),Microbiota = "GF"))
nrow(subset(subset(subset(sex, Sex == "M"), Population == "RS"),Microbiota = "CV"))
nrow(subset(subset(subset(sex, Sex == "M"), Population == "RS"),Microbiota = "GF"))

nrow(subset(subset(subset(sex, Sex == "F"), Population == "Bt"),Microbiota = "CV"))
nrow(subset(subset(subset(sex, Sex == "F"), Population == "Bt"),Microbiota = "GF"))
nrow(subset(subset(subset(sex, Sex == "F"), Population == "RS"),Microbiota = "CV"))
nrow(subset(subset(subset(sex, Sex == "F"), Population == "RS"),Microbiota = "GF"))

#It is clear that our samples are equivalent across treatment and only differ by population, so we will make that data tabular

Sex_Table = NULL
Male  = c(21,24)
Female = c(19,16)
Sex_Table = data.frame(Male,Female)
row.names(Sex_Table) = c("Bt","RS")
pander(Sex_Table)
21/19 #Proportion of 1.11 Males to every female for Bt population
24/16 #Proportion of 1.5 Males to every female for RS population
```
Sex represents an unbalanced factor since the samples are not equivalent in each factor. It is worth noting that there is a greater amount of males sampled from both populations and while the Bt sampling is almost equivalent there is a large discrepancy of 8 individuals between the RS female-Male sampling. Our treatment samples are all equivalent, unbalance is only observed in Population.

## Survival

Here we will examine if there is a significant relationship between the population of origin/microbiota and surival of the specimen. We have 160 observations, this will be done using a logistic regression model.

```{r}
#First we make a table one looking at population as the explanatory variable and one looking at microbiota
nrow(subset(subset(survival, Population == "RS"), Survived == "Yes"))
nrow(subset(subset(survival, Population == "Bt"), Survived == "Yes"))
nrow(subset(subset(survival, Population == "RS"), Survived == "No"))
nrow(subset(subset(survival, Population == "Bt"), Survived == "No"))

RS = c(61,19)
Bt = c(63,17)
survival_pop = data.frame(RS, Bt)
row.names(survival_pop) = c("Yes", "No")
survival_pop = t(survival_pop)
survive = as.factor(c(1,0))

survival_pop_glm = glm(survival_pop~survive, family = binomial(link = logit))
#summary(survival_pop_glm)
#anova(survival_pop_glm)

#Generating a logistic regression model treating our survival data as binomial (binary logit model) we see that population does not appear to be significant as a predictor of survival (p-value = .705)


#Now we'll look at Microbiota

nrow(subset(subset(survival, Microbiota == "CV"), Survived == "Yes"))
nrow(subset(subset(survival, Microbiota == "GF"), Survived == "Yes"))
nrow(subset(subset(survival, Microbiota == "CV"), Survived == "No"))
nrow(subset(subset(survival, Microbiota == "GF"), Survived == "No"))

CV = c(56,24)
GF = c(68,12)
survival_micro = data.frame(CV, GF)
row.names(survival_micro) = c("Yes", "No")
survival_micro = t(survival_micro)
survive = as.factor(c(1,0))

survival_micro_glm = glm(survival_micro~survive, family = binomial(link = logit))
#summary(survival_micro_glm)
#anova(survival_micro_glm)

#Finally make a logit model which looks at both factors simultaneously
mylogit <- glm(Survived ~ Population + Microbiota, data = survival, family = "binomial")
summary(mylogit)
anova(mylogit)
```

A logistic regression was performed on the survival data which found that population was not significant on survival rate, p-value of 0.7, however the microbiota treatment had a significant effect, p-value = 0.025.