{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Quality Analysis of Data\n",
    "\n",
    "## FastQC\n",
    "\n",
    "### Gularis\n",
    "\n",
    "First the files were unzipped into our shared BGMP directory:\n",
    "```\n",
    "jar -xf /home/lnassar/Project/RNA_Seq_data/gularis_RNASeq.zip\n",
    "jar -xf /home/lnassar/Project/RNA_Seq_data/inornata_RNASeq.zip\n",
    "```\n",
    "    Note: unzip did not work in this instance as the files were too large\n",
    "    \n",
    "The gularis directory had two separate runs, with each of them containing 28 files. The 2086 was subjected to a ribosomal depletion an the 2088 was a poly A selection. Also worth noting is the inclusion of a samples.txt file with information on the run/samples including tissue of origin. Of those 28 files half were R1, half R2. Additionally they were broken down into what seems the first half of a read, and the second. So the S_N_N nomenclature means:\n",
    "\n",
    "S_**1**_1_Index = The bold '1' is the first part of the read\n",
    "\n",
    "S_1_**1**_Index = The bold '1' here signifies it is the forward strand\n",
    "\n",
    "So 28/4 = There are 7 unique stranded reads in each of the two runs.\n",
    "\n",
    "A slurm script was made using FastQC on each individual file and outputing it to a unique folder. Following are the non-header parts of the scripts\n",
    "```\n",
    "#For Gularis 2036 (this is one of the two sequence identifiers for the runs, the other is 2038)\n",
    "ml fastqc\n",
    "fastqc -o /projects/bgmp/lizards/Gularis_RNAseq/gularis_RNASeq/MOLNG2036/FastQC_Gularis_2036 *.fastq.gz\n",
    "```\n",
    "```\n",
    "#For Gularis 2038\n",
    "ml fastqc\n",
    "fastqc -o /projects/bgmp/lizards/Gularis_RNAseq/gularis_RNASeq/MOLNG2038/FastQC_Gularis_2038 *.fastq.gz\n",
    "```\n",
    "    Note: Version 0.11.5 of FastQC was used for this analyses\n",
    "\n",
    "### Inornata\n",
    "\n",
    "Similar to gularis first the files were unzipped into the afformentioned directory. In this directory all of it was done in a single run, so you have half the data as the gularis data set. The data however is broken into 32 files instead of 28, so instead of 7 unique stranded reads for inornata we have 8 unique stranded reads. The titles are also easier to read with R1/R2 denotation and origin of the tissue.\n",
    "\n",
    "Like Gularis, a slurm script was submitted. The same FastQC version as previous was used. The non-header was as follows:\n",
    "```\n",
    "ml fastqc\n",
    "fastqc -o /projects/bgmp/lizards/Inornata_RNAseq/inornata_RNASeq/FastQC_Inornata *.fastq.gz\n",
    "```\n",
    "\n",
    "## MultiQC\n",
    "\n",
    "MultiQC is a software that aggregates results from bioinformatics analyses across many samples into a single report. In our case it is being used to combine all of the FastQC reports from each respective sequencing run. First it was installed with the following commands:\n",
    "\n",
    "```\n",
    "git clone https://github.com/ewels/MultiQC.git\n",
    "cd MultiQC\n",
    "python setup.py install\n",
    "```\n",
    "\n",
    "After it was installed all the FastQC files were copied to the local directory and from within each directory MultiQC was run with the simple command:\n",
    "\n",
    "```\n",
    "multiqc .\n",
    "```\n",
    "    Note: Version used was v1.3.dev0\n",
    "    \n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
