{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total length  of the genome accross all contigs is 857961\n",
      "The mean contig value is 2344.155737704918\n",
      "The largest contig length is 45994\n",
      "The mean coverage value is 424.81736178781875\n",
      "The N50 value for this assembly is 714\n",
      "There total number of contigs present in the file is 366\n",
      "# Contig length \t Number of contigs in category\n",
      "100\t15\n",
      "200\t69\n",
      "300\t43\n",
      "400\t19\n",
      "500\t16\n",
      "600\t17\n",
      "700\t16\n",
      "800\t10\n",
      "900\t9\n",
      "1000\t4\n",
      "1100\t8\n",
      "1200\t7\n",
      "1300\t4\n",
      "1400\t4\n",
      "1500\t8\n",
      "1700\t1\n",
      "1800\t3\n",
      "1900\t3\n",
      "2000\t4\n",
      "2100\t5\n",
      "2200\t4\n",
      "2300\t1\n",
      "2400\t2\n",
      "2500\t2\n",
      "2600\t1\n",
      "2700\t6\n",
      "2800\t1\n",
      "2900\t5\n",
      "3000\t1\n",
      "3100\t1\n",
      "3200\t2\n",
      "3300\t3\n",
      "3400\t2\n",
      "3500\t4\n",
      "3600\t2\n",
      "3700\t1\n",
      "3800\t1\n",
      "3900\t3\n",
      "4000\t1\n",
      "4200\t2\n",
      "4300\t2\n",
      "4400\t2\n",
      "4600\t2\n",
      "4700\t1\n",
      "4800\t1\n",
      "4900\t3\n",
      "5000\t1\n",
      "5100\t4\n",
      "5300\t1\n",
      "5400\t1\n",
      "5600\t1\n",
      "6100\t1\n",
      "6200\t1\n",
      "6500\t1\n",
      "6800\t5\n",
      "7000\t1\n",
      "7300\t1\n",
      "7700\t2\n",
      "7800\t1\n",
      "7900\t1\n",
      "8400\t1\n",
      "8600\t1\n",
      "9100\t1\n",
      "9200\t1\n",
      "9300\t1\n",
      "9800\t1\n",
      "10600\t1\n",
      "10900\t2\n",
      "11400\t1\n",
      "12100\t1\n",
      "13700\t1\n",
      "14200\t1\n",
      "14300\t1\n",
      "14400\t1\n",
      "15100\t1\n",
      "16800\t1\n",
      "18300\t1\n",
      "20100\t1\n",
      "22500\t1\n",
      "22600\t1\n",
      "26000\t1\n",
      "45900\t1\n"
     ]
    }
   ],
   "source": [
    "##PS6 Part 1\n",
    "my_file = open(\"contigs.fa\")\n",
    "#Specify your used kmer\n",
    "kmer = 31\n",
    "contigs = []\n",
    "kmer_length = []\n",
    "kmer_coverage = []\n",
    "total_contigs = 0\n",
    "for each in my_file:\n",
    "    if each.startswith(\">\"):\n",
    "        total_contigs = total_contigs + 1\n",
    "        contigs.append(each.strip('\\n'))\n",
    "\n",
    "#Isolate the kmer counts and depth coverage and convert to sequence length\n",
    "for each in contigs:\n",
    "    current=each.split(\"_\")\n",
    "    ###Append the converted kmer length with a kmer of 49, Kcount=L-K+1, in the case of #1: 3317=L-49+1\n",
    "    kmer_length.append(int(current[3])+(kmer-1))\n",
    "    kmer_coverage.append((float(current[5])*(float(current[3])+(kmer-1))/((float(current[3])+(kmer-1))-(kmer-1))))\n",
    "    \n",
    "#Caculate the total length of genome accross all contigs\n",
    "total_genome = sum(kmer_length)\n",
    "print(\"The total length  of the genome accross all contigs is\",total_genome)\n",
    "\n",
    "#Find the mean contig\n",
    "mean_contig=sum(kmer_length) / len(kmer_length)\n",
    "print(\"The mean contig value is\",mean_contig)\n",
    "\n",
    "#sort the kmer length list and print the largest value\n",
    "kmer_length = sorted(kmer_length)\n",
    "print(\"The largest contig length is\",kmer_length[total_contigs-1])\n",
    "\n",
    "#Calculate the mean depth of coverage\n",
    "mean_coverage=sum(kmer_coverage) / len(kmer_coverage)\n",
    "print(\"The mean coverage value is\",mean_coverage)\n",
    "\n",
    "#Calculate the N50 value, or median value\n",
    "print(\"The N50 value for this assembly is\",kmer_length[int(total_contigs/2)])\n",
    "\n",
    "#print(contigs)\n",
    "print(\"There total number of contigs present in the file is\",total_contigs)\n",
    "\n",
    "#Bin the kmer lengths using floor division and a dictionary\n",
    "binval = {}\n",
    "for each in kmer_length:\n",
    "    if each//100 not in binval:\n",
    "        binval[each//100]=1\n",
    "    else:\n",
    "        binval[each//100]=binval[each//100]+1\n",
    "\n",
    "#Print my binned contigs\n",
    "print (\"# Contig length \\t Number of contigs in category\")\n",
    "for keys,values in binval.items():\n",
    "    print(str(int(keys)*100)+\"\\t\"+str(values))\n",
    "my_file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
