{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "#Set my file variables and open by dictionary\n",
    "#tablefile=open(\"zebrafish_table.txt.txt\")\n",
    "tablefile=open(\"human_table.txt\")\n",
    "#proteinfile=open(\"Danio_rerio.GRCz10.pep.all.fa\")\n",
    "proteinfile=open(\"Homo_sapiens.GRCh38.pep.all.fa\")\n",
    "tableIDs = {}\n",
    "\n",
    "#Test file for QA\n",
    "#proteinfile=open(\"test.txt\")\n",
    "\n",
    "#Populate my dictionary keys with the geneIDs that also match for having a proteinID\n",
    "for lines in tablefile:\n",
    "    lines=lines.split(\"\\t\")\n",
    "    if len(lines[2]) != 1 and len(lines[0]) != 14:\n",
    "        tableIDs[lines[0].strip(\"\\n\")] = []\n",
    "\n",
    "#Iterate through the file and make it into a list\n",
    "fasta = []\n",
    "for lines in proteinfile:\n",
    "   lines = lines.strip(\"\\n\")\n",
    "   fasta.append(lines)\n",
    "\n",
    "#Iterate through the list of the file and match gene IDs to the dictionary and find\n",
    "#and save the sequences, replacing them and keeping only the longest\n",
    "n=-1\n",
    "for each in fasta:\n",
    "    n+=1\n",
    "#These first 2 conditions are to account for the first line\n",
    "####THE GENE AND PROTEIN IDS NEED TO BE CHANGED BETWEEN FISH AND HUMANS\n",
    "    if n==0:\n",
    "        header=each\n",
    "        #proteinID=header[1:19] #TURN THIS ON FOR FISH PROTEOME\n",
    "        proteinID=header[1:16] #TURN THIS ON FOR HUMAN PROTEOME\n",
    "        each=each.split(\" \")\n",
    "        #geneID = each[3][5:23] #TURN THIS ON FOR FISH PROTEOME\n",
    "        geneID = each[3][5:20] #TURN THIS ON FOR HUMAN PROTEOME\n",
    "    elif n==1:\n",
    "        seq=each\n",
    "    elif re.search(\"^>\",each):\n",
    "        if geneID in tableIDs.keys() and tableIDs[geneID]==[]:\n",
    "            tableIDs[geneID].append(header)\n",
    "            tableIDs[geneID].append(seq)\n",
    "        elif len(seq) > len(tableIDs[geneID][1]):\n",
    "            tableIDs[geneID][0] = header\n",
    "            tableIDs[geneID][1] = seq\n",
    "        header = each\n",
    "        #proteinID=header[1:19] #TURN THIS ON FOR FISH PROTEOME\n",
    "        proteinID=header[1:16] #TURN THIS ON FOR HUMAN PROTEOME\n",
    "        each=each.split(\" \")\n",
    "        #geneID = each[3][5:23] #TURN THIS ON FOR FISH PROTEOME\n",
    "        geneID = each[3][5:20] #TURN THIS ON FOR HUMAN PROTEOME\n",
    "        seq = \"\"\n",
    "#Following loop is to account for the last line in the file\n",
    "    elif len(fasta) == n+1:\n",
    "        seq = seq + each\n",
    "        if geneID in tableIDs.keys() and tableIDs[geneID]==[]:\n",
    "            tableIDs[geneID].append(header)\n",
    "            tableIDs[geneID].append(seq)\n",
    "        elif len(seq) > len(tableIDs[geneID][1]):\n",
    "            tableIDs[geneID][0] = header\n",
    "            tableIDs[geneID][1] = seq\n",
    "\n",
    "#Add the sequence lines\n",
    "    else:\n",
    "        seq = seq + each\n",
    "\n",
    "#Output our fasta format\n",
    "for keys, values in tableIDs.items():\n",
    "   print(tableIDs[keys][0]+\"\\n\"+tableIDs[keys][1])\n",
    "\n",
    "#QA and file closing\n",
    "#print(\"the number of lines in the file is\",n)\n",
    "#print(\"The length of your fasta is\",len(fasta))\n",
    "#print(tableIDs)\n",
    "tablefile.close()\n",
    "proteinfile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
